### PR TITLE
Add implementation

### PR DESCRIPTION
**Pull request by AdnoC**

### CODE REVIEW

File: src/axePuppeteer.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+      throw new Error(
+        'AxePuppeteer needs an object to configure. See axe-core configure API.'
+      )
+    }
+
+    this.config = config
+    return this
+  }
+
+  // TODO: Confirm that we don't have to resolve promise if given promise.
```
[the remark this answers is not in the document]
Comment (AdnoC) [reply src/axePuppeteer.ts]: I thought I removed that comment. My bad.

File: test/utils.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+export function fixtureFilePath(filename: string): string {
+  return path.resolve(__dirname, 'fixtures', filename)
+}
+
+export async function customConfig() {
+  const configFile = fixtureFilePath('custom-rule-config.json')
+  const config = JSON.parse(
+    await fs.readFile(configFile, 'utf8')
```
[the remark this answers is not in the document]
Comment (AdnoC) [reply test/utils.ts]: Unless you feel strongly that it should be changed to a synchronous read, I'm going to keep it async.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,78 @@
+{
+  "name": "axe-puppeteer",
+  "version": "0.0.1",
+  "description": "Provides a chainable axe API for Puppeteer and automatically injects into all frames",
+  "main": "dist/index.js",
+  "typings": "dist/index.d.ts",
+  "scripts": {
+    "build": "tsc",
+    "test": "mocha test/**/*.ts src/**/*.ts",
```
[the remark this answers is not in the document]
Comment (AdnoC) [reply package.json]: That was something done while trying to get coverage to work. Not needed anymore.

File: src/axePuppeteer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdnoC) [reply src/axePuppeteer.ts]: Made the change but I can't push.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      - restore_cache:
+          key: v1-npm-cache-{{ checksum "package.json" }}
+      - run: npm run lint
+  release:
+    <<: *defaults
+    steps:
+      - checkout
+      - restore_cache:
+          key: v1-npm-cache-{{ checksum "package.json" }}
+      - run: npm config set "//registry.npmjs.org/:_authToken" $NPM_AUTH
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply .circleci/config.yml]: I added the topic for the same user we use to publish axe-core from CI.

File: src/axePuppeteer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply src/axePuppeteer.ts]: Why can't you push? What does that mean?

File: test/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply test/utils.ts]: Nah, not a big deal. This is fine 👍

File: src/axePuppeteer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdnoC) [reply src/axePuppeteer.ts]: That means that when I do `git push origin develop` I get the error message

```
ERROR: Permission to dequelabs/axe-puppeteer.git denied to AdnoC.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

File: src/owning.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+export async function loadPage(
+  browser: Browser,
+  url: string,
+  { opts, source }: { opts?: any; source?: string } = {}
+) {
+  const page = await browser.newPage()
+  await page.setBypassCSP(true)
+
+  await page.goto(url, opts)
```
Comment (jkodu) [inline src/owning.ts]: In my experience with puppeteer, it is best to wrap these fns by a try/catch block. Helps debugging a lot. Worth doing it across all usages.

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,64 @@
+{
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply tsconfig.json]: I agree here, but it didn't seem worth blocking this from landing.

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      - checkout
+      - restore_cache:
+          key: v2-npm-cache-{{ checksum "package.json" }}
+      - run: npm run lint
+  release:
+    <<: *defaults
+    steps:
+      - checkout
+      - restore_cache:
+          key: v2-npm-cache-{{ checksum "package.json" }}
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply .circleci/config.yml]: I agree with "v1". I do no think it's worth keeping the cache version in a variable tho.

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  build:
+    jobs:
+      - dependencies
+      - tests:
+          requires:
+            - dependencies
+      - lint:
+          requires:
+            - dependencies
+      - release:
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply .circleci/config.yml]: Agreed and commented on the other PR :)

File: tsconfig.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (AdnoC) [reply tsconfig.json]: The file was generated by `tsc --init`. I just uncommented the lines of options to enable (and changed their values). Can certainly remove the unused options.

File: .circleci/config.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (AdnoC) [reply .circleci/config.yml]: I believe I had a reason to change it to `v2`. Something refused to update so I changed the cache. Can change it and see if it works.

File: tslint.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+  "defaultSeverity": "error",
+  "extends": [
+    "tslint:recommended",
+    "tslint-config-standard",
+    "tslint-config-prettier"
+  ],
+  "jsRules": {},
```
[the remark this answers is not in the document]
Comment (AdnoC) [reply tslint.json]: I just took this from one of our other Typescript projects.

File: src/browser.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+// This module encapsulates the browser enviromnent
+import Axe from 'axe-core'
+
+// Expect axe to be set up.
+declare var axe: typeof Axe
```
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply src/browser.ts]: I strongly prefer `window.axe`.

File: src/browser.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+// Expect axe to be set up.
+// Tell Typescript that there should be a variable called `axe` that follows
+// the shape given by the `axe-core` typings (the `run` and `configure` functions).
+declare var axe: typeof Axe
+// Defined at top-level to clarify that it can't capture variables from outer scope.
+export function runAxe(
+  config: Axe.Spec | null,
+  context: Axe.ElementContext | null,
+  options: Axe.RunOptions | null
```
[the remark this answers is not in the document]
Comment (AdnoC) [reply src/browser.ts]: Sure thing

File: src/owning.ts (hunk 5 above)
Comment (AdnoC) [reply src/owning.ts]: What should the `catch` block do? Rethrow with a different message? Close the new page?

File: .circleci/config.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkodu) [reply .circleci/config.yml]: Yet another reason, why it would save us a commit, if it was an env variable 😉

File: src/owning.ts (hunk 5 above)
Comment (jkodu) [reply src/owning.ts]: Up to you, throw the same error, but with better diagnostics, as you please. Again, not a big thing.

**Pull request closed without merging**